enhancements/api-review/custom-tags-aws.md

```diff
@@ -79,65 +99,323 @@ The userTags field is intended to be set at install time and is considered immut
+   Action = No change in tags for the AWS resource.\
+   Final tag set to AWS resource = `key_infra1 = value1`\
+   Event action = No event
+
+3) User sets user-defined tag created using external tools to empty string for deleting for AWS resource.\
+   There is no validation check involved for creator tool of user-defined tag. Any user-defined tag listed by user in `.spec.platformSpec.aws.resourceTags` is considered for create/update/delete accordingly.
+
+4) Any user-defined tag set using `.spec.platformSpec.aws.resourceTags` in `Infrastructure.config.openshift.io/v1` type has scope limited to cluster-level.
+
+5) User-defined tags are not synced from `.spec.platformSpec.aws.resourceTags` to `.status.platformStatus.aws.resourceTags` for the following reasons.
+- User-defined tags in `.spec.platformSpec.aws.resourceTags` can be "removed without delete", "delete", "update".
+  `.spec.platformSpec.aws.resourceTags` to `.status.platformStatus.aws.resourceTags` sync is not required,
+  as there will be no versioning of the user-defined tags required to override the user-defined tags in Infrastructure CR.
+  Instead, the user-defined tags (if supported in resource operator spec field, e.g: machine) can override the user-defined tags in Infrastructure CR.
+- As `.spec.platformSpec.aws.resourceTags` has the actual values. Sync to `.status.platformStatus.aws.resourceTags` was required in earlier design to identify if the
+  tag on the resource was created using Infrastructure CR or external tool. Identifying the creator tool was done to restrict user from editing the same user-tag kv pair using multiple tools.
+  This inherently poses many scenarios of conflict which will result in user-defined tag kv pair being inconsistent across cluster when applied using Infrastructure CR. Hence, user will be confused which tool to be used to update tag.
+
+6) User applies user-defined tag using `.spec.platformSpec.aws.resourceTags`. Later, user modifies the user-defined tag for AWS resource using external tool.
+   In this case, the desired value is not set immediately for AWS resource by the owning operator. There is eventual consistency maintained by the owning operator.
+   The time taken to reconcile the modified user-defined tag on AWS resource to desired value vary across owning operators.
+
+7) Post-installation of the cluster, validation for kubernetes.io or openshift.io namespaced tags in the `.spec.platformSpec.aws.resourceTags` is not done by API server.
+   The individual AWS resource component operator fails to apply the user tags. Developing a production-ready webhook is a substantial task. To support webhook life-cycle management and monitoring, instrumentation, alerting and integration with the packaging/releases processes must be supported.
+   For validating fixed strings it is better to have an expression based check done by API server which is already part of upstream as alpha feature for CRD Validation Expression Language.
```

Comment:
It is better to have validation check at api server to avoid stale entries in spec field and additional triggers for removing it.
As user tags can be updated without validation on AWS direclty, to raise a warning or overwrite, there needs to be a validation check at the operator.
Kube resource objects are not validated and allow namespace strings (`kubernetes.io` and `openshift.io`) to be used in user tags. Operators need to validate to avoid breaking change.

As the webhook implementation might become obsolete in next versions, allowing stale entries in spec field for now, does not allow invalid usage of namespace strings.

Reply:
Given that is seems that there will be no "one size fits all" logic to how the tags will be applied (e.g. create vs update and resource specific limitations) I don't think it is unreasonable to simply have each operator also validate the keys names while they are at it. A helper can be provided in library go to help with this.

Reply:
Another option would be for the config-controller to copy only valid resourceTag entries into the `infrastructure.status.platformStatus.aws.resourceTags` and have operators observe those values instead of the `platformSpec.aws.resourceTags`.

Reply:
I'm of the strong opinion that if we ever want validation, it needs to be validated centrally and from the beginning of the introduction of the feature (we had these limitations on the previous iteration, we shouldn't lose them now).

While a webhook introduced now may be obsoleted in the future, the alpha feature may never graduate. I see this as a blocker for this feature and I believe the API reviewers will block the API implementation without this

Is there an API PR created already?

Reply:
To clarify some points here, I don't disagree with what @sanchezl is saying. We have a number of different scenarios to work with here.

1) Tags in the platformStatus - these are validated by the installer to ensure that they meet the namespace requirements, therefore any operator can apply them blindly without validation
2) Tags on a local object (eg Machine) - these must be validated by the operator when they are merged with the platformSpec and platformStatus tags, however, not every operator has these. 
3) Tags on the platformSpec
  a) If these are validated centrally, any operator applying them can apply them blindly, therefore, if they don't have their own local tags, they need to do no validation and can blindly apply both platformSpec and platformStatus tags. If they don't have their own local tags, they probably don't have a way to report that the tags are invalid (I would expect those with this ability to validate via a webhook)
  b) If these aren't validated centrally, each operator must validate them and reject anything that is invalid, this is async and so now we are left with the problem of, how is this reported to the user? Some operators will have status objects they put this on, some may not. What happens if the only reporting mechanism an operator has currently is via events? The UX of this, because its async, will not be as good as option a

Reply:
After discussion, the validation the is considered is default validation method supported : https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#defaulting